win32/win32.c

```diff
@@ -2615,9 +2615,73 @@ set_pioinfo_extra(void)
      * * https://bugs.ruby-lang.org/issues/18605
      */
     char *p = (char*)get_proc_address(UCRTBASE, "_isatty", NULL);
-    char *pend = p;
     /* _osfile(fh) & FDEV */
 
+#ifdef _M_ARM64
+#define IS_INSN(pc, name) ((*(pc) & name##_mask) == name##_id)
+    const int max_num_inst = 500;
+    uint32_t *start = (uint32_t*)p;
+    uint32_t *end_limit = (start + max_num_inst);
+    uint32_t *pc = start;
+
+    if (!p) {
+        fprintf(stderr, "_isatty proc not found in " UCRTBASE "\n");
+        _exit(1);
+    }
+
+    /* end of function */
+    const uint32_t ret_id = 0xd65f0000;
+    const uint32_t ret_mask = 0xfffffc1f;
+    for(; pc < end_limit; pc++) {
+        if (IS_INSN(pc, ret)) {
+            break;
+        }
+    }
+    if (pc == end_limit) {
+        fprintf(stderr, "end of _isatty not found in " UCRTBASE "\n");
+        _exit(1);
+    }
+
+    /* pioinfo instruction mark */
+    const uint32_t adrp_id = 0x90000000;
+    const uint32_t adrp_mask = 0x9f000000;
+    const uint32_t add_id = 0x11000000;
+    const uint32_t add_mask = 0x3fc00000;
+    for(; pc > start; pc--) {
+        if (IS_INSN(pc, adrp) && IS_INSN(pc + 1, add)) {
+            break;
+        }
+    }
+    if(pc == start) {
+        fprintf(stderr, "pioinfo mark not found in " UCRTBASE "\n");
+        _exit(1);
+    }
+
+    /* We now point to instructions that load address of __pioinfo:
+     * adrp	x8, 0x1801d8000
+     * add	x8, x8, #0xdb0
+     * https://devblogs.microsoft.com/oldnewthing/20220809-00/?p=106955
+     * The last adrp/add sequence before ret is what we are looking for.
+     */
+    const uint32_t adrp_insn = *pc;
+    const uint32_t adrp_immhi = (adrp_insn & 0x00ffffe0) >> 5;
+    const uint32_t adrp_immlo = (adrp_insn & 0x60000000) >> (5 + 19 + 5);
+    /* imm = immhi:immlo:Zeros(12), 64 */
+    const uint64_t adrp_imm = ((adrp_immhi << 2) | adrp_immlo) << 12;
```

Comment:
I'm messing with some similar code, and noticed what I think is a bug here: adrp's immediate should be sign-extended.  I think something like 
```suggestion
    /* imm = SignExtend(immhi:immlo:Zeros(12), 64) */
    int64_t adrp_imm = (adrp_immhi << 2) | adrp_immlo;
    /* sign extend */
    if (adrp_imm & (1 << 20))
        adrp_imm |= ~((1 << 21) - 1);
    adrp_imm <<= 12;
```

if you really want that to be `const` you could do that in a ternary but I don't know if it's more legible that way:
```suggestion
    /* imm = SignExtend(immhi:immlo:Zeros(12), 64) */
    const int64_t adrp_imm = (((adrp_immhi & (1 << 18)) ? (1 << 21) - 1 : 0) | (adrp_immhi << 2) | adrp_immlo) << 12;
```

Reply:
You're right, I missed that, and luckily, we didn't it the case where adrp immhi sign matters.
I won't push a fix, but you're welcome to contribute it.

Reply:
Maybe... The code I'm messing with didn't hit a negative offset either, but I was thinking it wouldn't be very useful if it wasn't signed, so checked the docs instead of just relying on what this code was doing... I didn't check the specifics of the add instruction, the code I was looking at didn't use that, instead an `ldr` with an immediate unsigned offset to add in the low-order bits.

Reply:
By curiosity, I checked the QEMU source as well, and sign is carefully extended for the immediate when decoding instruction.